diffrax/_step_size_controller/jump_step_wrapper.py

```diff
@@ -0,0 +1,424 @@
+from collections.abc import Callable
+from typing import Generic, Optional, TYPE_CHECKING, TypeVar
+
+import equinox as eqx
+import equinox.internal as eqxi
+import jax
+import jax.numpy as jnp
+from jaxtyping import Array, PyTree, Real
+
+from .._custom_types import (
+    Args,
+    BoolScalarLike,
+    IntScalarLike,
+    RealScalarLike,
+    VF,
+    Y,
+)
+from .._misc import static_select, upcast_or_raise
+from .._solution import RESULTS
+from .._term import AbstractTerm
+from .adaptive_base import AbstractAdaptiveStepSizeController
+from .base import AbstractStepSizeController
+
+
+_ControllerState = TypeVar("_ControllerState")
+_Dt0 = TypeVar("_Dt0", None, RealScalarLike, Optional[RealScalarLike])
+
+
+class _JumpStepState(eqx.Module, Generic[_ControllerState]):
+    made_jump: BoolScalarLike
+    prev_dt: RealScalarLike
+    step_index: IntScalarLike
+    jump_index: IntScalarLike
+    rejected_index: IntScalarLike
+    rejected_buffer: Optional[Array]
+    step_ts: Optional[Array]
+    jump_ts: Optional[Array]
+    inner_state: _ControllerState
+
+    def get(self):
+        return (
+            self.made_jump,
+            self.prev_dt,
+            self.step_index,
+            self.jump_index,
+            self.rejected_index,
+            self.rejected_buffer,
+            self.step_ts,
+            self.jump_ts,
+            self.inner_state,
+        )
+
+
+def _none_or_array(x):
+    if x is None:
+        return None
+    else:
+        return jnp.asarray(x)
+
+
+def _get_t(i: IntScalarLike, ts: Array) -> RealScalarLike:
+    i_min_len = jnp.minimum(i, len(ts) - 1)
+    return jnp.where(i == len(ts), jnp.inf, ts[i_min_len])
```

Comment:
For each of the tests (except backprop) in `test_adaptive_stepsize_controller.py` I added `@pytest.mark.parametrize("backwards", [False, True])` which just swaps the order of `t0` and `t1`. This also revealed two other problems:

1.  `step_ts` and `jump_ts` need to be sorted and in particular need to be re-sorted after multiplying them with `direction` in `wrap`. I added sorting both in `JSW.__init__` and `JSW.wrap`.
2. VBT complains if `t0 >= t1`. I don't think this is necessarily a problem but it could be confusing to some users, so let me know if you want to revisit that design decision and I can try to do something about it.

But other than that it all seems to work perfectly when solving backwards. To be fair the way you dealt with the whole backwards solve business is very clean and makes this work perfectly without any alterations.